Add the QuickBooks provider

yongshui opened the pull request:
Add the QuickBooks OAuth2.0 Support

Please refer to the following QuickBooks Authentication and authorization document URL:

https://www.developer.intuit.com/app/developer/qbo/docs/develop/authentication-and-authorization/oauth-2.0-playground

yongshui committed:
Add the QuickBooks provider

yongshui committed:
Change the author of QuickBooks property

yongshui committed:
Change QuickBooks project property from TargetFramework to TargetFrameworks

martincostello commented on AspNet.Security.OAuth.Providers.sln:
```diff
@@ -1,7 +1,7 @@
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28803.156
+# Visual Studio Version 17
+VisualStudioVersion = 17.0.31612.314
```
Please revert this. We don't want the solution to need Visual Studio 2022 until it's stable.

yongshui replied:
OK, I will revert this configuration.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/AspNet.Security.OAuth.QuickBooks.csproj:
```diff
@@ -0,0 +1,20 @@
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="IppDotNetSdkForQuickBooksApiV3" Version="14.6.1" />
```
What is this reference used for? We generally don't depend on third-party assemblies in the core provider code.

yongshui replied:
This third-party assembly will be removed.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/AspNet.Security.OAuth.QuickBooks.csproj:
```diff
@@ -0,0 +1,20 @@
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="IppDotNetSdkForQuickBooksApiV3" Version="14.6.1" />
```
Versions need to be defined in `Packages.props`.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/Packages.props

yongshui replied:
Moved the versions configuration from cjproj to packages.props

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationDefaults.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public const string UserInformationEndpoint = "https://sandbox-accounts.platform.intuit.com/v1/openid_connect/userinfo";
```
Sandbox?

yongshui replied:
Sandbox is for the Quickbooks development. I will add the server URL of the production environment in the readme file.

martincostello replied:
It doesn't seem particularly useful to have to reconfigure the provider (and read the documentation) for the default use case of using the production system.

Other providers in this repo provide constants/functionality for switching between environments, but the default should be the real system, not the test one.

yongshui replied:
The default provider URL changed to production.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+    /// </summary>
+    public class QuickBooksAuthenticationOptions : OAuthOptions
+    {
+        // private string scopeValC2QB = "com.intuit.quickbooks.accounting";
+        // private string scopeValSIWI = "com.intuit.quickbooks.payment";
```
Delete?

yongshui replied:
It doesn't need to delete the other scope options. I will enable the other scope options.

yongshui replied:
Removed the unnecessary scopes.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        // private string scopeValC2QB = "com.intuit.quickbooks.accounting";
+        // private string scopeValSIWI = "com.intuit.quickbooks.payment";
+        private string scopeValOpenId = "openid email profile phone address";
```
This doesn't seem right - claims are usually added individually, and then joined in the handler by `FormatScope()`.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+            TokenEndpoint = QuickBooksAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = QuickBooksAuthenticationDefaults.UserInformationEndpoint;
+
+            Scope.Add(scopeValOpenId);
```
I would expect this to be:

```suggestion
            Scope.Add("openid");
            Scope.Add("email");
            Scope.Add("profile");
            Scope.Add("phone");
            Scope.Add("address");
```

yongshui replied:
Unfortunately, your suggestion codes of the scope are not valid. The server will throw an invalid scope exception.

martincostello replied:
Sounds to me like a bug of `FormatScopes()` not using the right delimiter. The scopes should be a list, and then the handler should join them together in a way that makes the service accept them.

yongshui replied:
Changed to the recommended codes.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        // private string scopeValC2QB = "com.intuit.quickbooks.accounting";
+        // private string scopeValSIWI = "com.intuit.quickbooks.payment";
+        private string scopeValOpenId = "openid email profile phone address";
```
Are all these scopes _required_? In general we only add the minimum required scopes for the basic functionality - the user can configure more scopes and claims as needed to extend the functionality.

yongshui replied:
The QuickBooks defines three categories scope. and the openId category includes 5 sub-scopes. And it looks like all 5 sub-scopes can not be split.

yongshui replied:
I've removed other unnecessary scope configurations and keep the minimum required scope.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj:
```diff
@@ -17,6 +17,10 @@
     <ProjectReference Include="..\..\src\AspNet.Security.OAuth.*\*.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="QuickBooks\bundle.json" />
+  </ItemGroup>
+
```
```suggestion
```

yongshui replied:
Why do you suggest deleting this test configuration? It seems it's necessary for the xunit test.

martincostello replied:
It is not necessary. All that's needed is what's on line 12.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/9c347a6d3ee06b6801cdc0a083549b150ae3f00f/test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj#L12

yongshui replied:
Removed

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/QuickBooks/QuickBooksTests.cs:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.Email, "rekingshui@gmail.com")]
```
Please add a test case for each claim mapped by the default options.

yongshui replied:
OK, I will add all the claims mapped by the default options.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/QuickBooks/QuickBooksTests.cs:
```diff
@@ -0,0 +1,48 @@
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.Email, "rekingshui@gmail.com")]
```
Is this your real email address? Do you really want it in our tests?

yongshui replied:
I will change to a fake email address. although the real email address is not big deal.

yongshui committed:
Update AspNet.Security.OAuth.Providers.sln

yongshui committed:
Update AspNet.Security.OAuth.QuickBooks.csproj

yongshui committed:
Changed the QuickBooks authentication options.
Add the test cases for all the mapped claims.
Change the email to a fake email address.

yongshui committed:
Merge branch 'dev' of https://github.com/yongshui/AspNet.Security.OAuth.Providers into dev

yongshui committed:
Move the version from the csproj to pakcages.props.
Clean the QuickBooksAuthenticationHandler.

martincostello commented on Packages.props:
```diff
@@ -13,6 +13,7 @@
     <PackageReference Update="Shouldly" Version="4.0.3" />
     <PackageReference Update="StyleCop.Analyzers" Version="1.1.118" />
     <PackageReference Update="System.IdentityModel.Tokens.Jwt" Version="6.6.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.OAuth" Version="2.2.0" />
```
This is redundant, the providers target ASP.NET Core 5 where this is built in.
```suggestion
```

yongshui replied:
Removed

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/AspNet.Security.OAuth.QuickBooks.csproj:
```diff
@@ -0,0 +1,18 @@
+    <Description>ASP.NET Core security middleware enabling QuickBooks authentication.</Description>
+    <Authors>RekingShui</Authors>
+    <PackageTags>quickbooks;aspnetcore;authentication;oauth;security</PackageTags>
+    <GeneratePackageOnBuild>false</GeneratePackageOnBuild>
```
```suggestion
```

yongshui replied:
Removed

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -0,0 +1,38 @@
+    /// </summary>
+    public class QuickBooksAuthenticationOptions : OAuthOptions
+    {
+        private readonly string scopeValOpenId = "openid email profile phone address";
```
It would seem to me that the _minimum_ required scopes are `openid` and `profile`. The rest are extras.

yongshui replied:
The email and phone scopes are necessary.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationHandler.cs:
```diff
@@ -0,0 +1,71 @@
+            return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+        }
+
+        /// <inheritdoc/>
+        protected override string FormatScope([NotNull] IEnumerable<string> scopes)
+            => string.Join(',', scopes);
```
If this override is removed, it will use the base class behaviour of delimiting with a space. Then the scopes will work as a list like our other providers.

```suggestion
```

yongshui replied:
removed

yongshui committed:
Removed the unnecessary configuration and codes
Changed the QuickBooks provider URL from test to production.
Changed the scope implementation as recommended.

martincostello commented on src/AspNet.Security.OAuth.QuickBooks/QuickBooksAuthenticationOptions.cs:
```diff
@@ -27,7 +25,10 @@ public QuickBooksAuthenticationOptions()
-            Scope.Add(scopeValOpenId);
+            Scope.Add("openid");
+            Scope.Add("profile");
+            Scope.Add("phone");
+            Scope.Add("email");
```
Are these two scopes _required_?

martincostello replied:
Sorry, just saw your reply.

martincostello commented:
Thanks! 🎉

This provider will ship with the next release. When that will be depends on one of the other open PRs, as I'd prefer to release two new providers together, rather than separately.

Until it's available from NuGet.org, you can use our prerelease feed in MyGet.

martincostello commented:
The Quickbooks provider will be available from [NuGet.org](https://www.nuget.org/packages/AspNet.Security.OAuth.QuickBooks/) shortly.

Thanks again for your contribution. 🙇
